File: src/main/java/kirjanpito/ui/javafx/dialogs/SettingsExportImportFX.java
package kirjanpito.ui.javafx.dialogs;

import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import javafx.stage.FileChooser;
import javafx.stage.Window;

import kirjanpito.util.AppSettings;

import java.io.*;
import java.util.Properties;
import java.util.Optional;

/**
 * Asetusten vienti ja tuonti.
 */
public class SettingsExportImportFX {
    
    private Window owner;
    private AppSettings settings;
    
    public SettingsExportImportFX(Window owner) {
        this.owner = owner;
        this.settings = AppSettings.getInstance();
    }
    
    /**
     * Vie asetukset tiedostoon.
     */
    public void exportSettings() {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Vie asetukset");
        fileChooser.getExtensionFilters().add(
            new FileChooser.ExtensionFilter("Asetustiedosto (*.properties)", "*.properties")
        );
        fileChooser.setInitialFileName("tilitin-asetukset.properties");
        
        File file = fileChooser.showSaveDialog(owner);
        if (file != null) {
            try {
                // Lue nykyiset asetukset
                File settingsFile = settings.getFile();
                if (settingsFile != null && settingsFile.exists()) {
                    Properties props = new Properties();
                    try (FileInputStream fis = new FileInputStream(settingsFile)) {
                        props.load(fis);
                    }
                    
                    // Kirjoita uuteen tiedostoon
                    try (FileOutputStream fos = new FileOutputStream(file)) {
                        props.store(fos, "Tilitin asetukset - viety " + java.time.LocalDateTime.now());
                    }
                    
                    showInfo("Asetukset viety", "Asetukset vietiin onnistuneesti:\n" + file.getAbsolutePath());
                } else {
                    showError("Virhe", "Asetustiedostoa ei löytynyt.");
                }
            } catch (IOException e) {
                showError("Virhe viennissä", "Asetusten vienti epäonnistui: " + e.getMessage());
            }
        }
    }
    
    /**
     * Tuo asetukset tiedostosta.
     */
    public void importSettings() {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Tuo asetukset");
        fileChooser.getExtensionFilters().add(
            new FileChooser.ExtensionFilter("Asetustiedosto (*.properties)", "*.properties")
        );
        
        File file = fileChooser.showOpenDialog(owner);
        if (file != null) {
            // Varmista käyttäjältä
            Alert confirm = new Alert(Alert.AlertType.CONFIRMATION);
            confirm.initOwner(owner);
            confirm.setTitle("Vahvista tuonti");
            confirm.setHeaderText("Tuodaanko asetukset?");
            confirm.setContentText("Nykyiset asetukset korvataan tuotavilla asetuksilla.\n\nHaluatko jatkaa?");
            
            Optional<ButtonType> result = confirm.showAndWait();
            if (result.isPresent() && result.get() == ButtonType.OK) {
                try {
                    // Lue tuotavat asetukset
                    Properties importProps = new Properties();
                    try (FileInputStream fis = new FileInputStream(file)) {
                        importProps.load(fis);
                    }
                    
                    // Päivitä asetukset
                    File settingsFile = settings.getFile();
                    if (settingsFile != null) {
                        try (FileOutputStream fos = new FileOutputStream(settingsFile)) {
                            importProps.store(fos, "Tilitin asetukset - tuotu " + java.time.LocalDateTime.now());
                        }
                        
                        // Lataa asetukset uudelleen
                        settings.load(settingsFile);
                        
                        showInfo("Asetukset tuotu", 
                            "Asetukset tuotiin onnistuneesti.\n\nOsa asetuksista astuu voimaan ohjelman uudelleenkäynnistyksen jälkeen.");
                    }
                } catch (IOException e) {
                    showError("Virhe tuonnissa", "Asetusten tuonti epäonnistui: " + e.getMessage());
                }
            }
        }
    }
    
    private void showInfo(String title, String content) {
        Alert alert = new Alert(Alert.AlertType.INFORMATION);
        alert.initOwner(owner);
        alert.setTitle(title);
        alert.setHeaderText(null);
        alert.setContentText(content);
        alert.showAndWait();
    }
    
    private void showError(String title, String content) {
        Alert alert = new Alert(Alert.AlertType.ERROR);
        alert.initOwner(owner);
        alert.setTitle(title);
        alert.setHeaderText(null);
        alert.setContentText(content);
        alert.showAndWait();
    }
}

File: src/main/kotlin/kirjanpito/ui/javafx/dialogs/SettingsExportImportFX.kt
package kirjanpito.ui.javafx.dialogs

import javafx.scene.control.Alert
import javafx.scene.control.ButtonType
import javafx.stage.FileChooser
import javafx.stage.Window
import kirjanpito.util.AppSettings
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.time.LocalDateTime
import java.util.Properties

/**
 * Asetusten vienti ja tuonti.
 */
class SettingsExportImportFX(private val owner: Window?) {
    
    private val settings: AppSettings = AppSettings.getInstance()
    
    /**
     * Vie asetukset tiedostoon.
     */
    fun exportSettings() {
        val fileChooser = FileChooser().apply {
            title = "Vie asetukset"
            extensionFilters.add(
                FileChooser.ExtensionFilter("Asetustiedosto (*.properties)", "*.properties")
            )
            initialFileName = "tilitin-asetukset.properties"
        }
        
        val file = fileChooser.showSaveDialog(owner)
        if (file != null) {
            try {
                // Lue nykyiset asetukset
                val settingsFile = settings.file
                if (settingsFile != null && settingsFile.exists()) {
                    val props = Properties()
                    FileInputStream(settingsFile).use { fis ->
                        props.load(fis)
                    }
                    
                    // Kirjoita uuteen tiedostoon
                    FileOutputStream(file).use { fos ->
                        props.store(fos, "Tilitin asetukset - viety ${LocalDateTime.now()}")
                    }
                    
                    showInfo("Asetukset viety", "Asetukset vietiin onnistuneesti:\n${file.absolutePath}")
                } else {
                    showError("Virhe", "Asetustiedostoa ei löytynyt.")
                }
            } catch (e: IOException) {
                showError("Virhe viennissä", "Asetusten vienti epäonnistui: ${e.message}")
            }
        }
    }
    
    /**
     * Tuo asetukset tiedostosta.
     */
    fun importSettings() {
        val fileChooser = FileChooser().apply {
            title = "Tuo asetukset"
            extensionFilters.add(
                FileChooser.ExtensionFilter("Asetustiedosto (*.properties)", "*.properties")
            )
        }
        
        val file = fileChooser.showOpenDialog(owner)
        if (file != null) {
            // Varmista käyttäjältä
            val confirm = Alert(Alert.AlertType.CONFIRMATION).apply {
                initOwner(this@SettingsExportImportFX.owner)
                title = "Vahvista tuonti"
                headerText = "Tuodaanko asetukset?"
                contentText = "Nykyiset asetukset korvataan tuotavilla asetuksilla.\n\nHaluatko jatkaa?"
            }
            
            val result = confirm.showAndWait()
            if (result.isPresent && result.get() == ButtonType.OK) {
                try {
                    // Lue tuotavat asetukset
                    val importProps = Properties()
                    FileInputStream(file).use { fis ->
                        importProps.load(fis)
                    }
                    
                    // Päivitä asetukset
                    val settingsFile = settings.file
                    if (settingsFile != null) {
                        FileOutputStream(settingsFile).use { fos ->
                            importProps.store(fos, "Tilitin asetukset - tuotu ${LocalDateTime.now()}")
                        }
                        
                        // Lataa asetukset uudelleen
                        settings.load(settingsFile)
                        
                        showInfo(
                            "Asetukset tuotu",
                            "Asetukset tuotiin onnistuneesti.\n\nOsa asetuksista astuu voimaan ohjelman uudelleenkäynnistyksen jälkeen."
                        )
                    }
                } catch (e: IOException) {
                    showError("Virhe tuonnissa", "Asetusten tuonti epäonnistui: ${e.message}")
                }
            }
        }
    }
    
    private fun showInfo(title: String, content: String) {
        Alert(Alert.AlertType.INFORMATION).apply {
            initOwner(this@SettingsExportImportFX.owner)
            this.title = title
            headerText = null
            contentText = content
        }.showAndWait()
    }
    
    private fun showError(title: String, content: String) {
        Alert(Alert.AlertType.ERROR).apply {
            initOwner(this@SettingsExportImportFX.owner)
            this.title = title
            headerText = null
            contentText = content
        }.showAndWait()
    }
}
